src/index.js

```diff
@@ -152,86 +155,93 @@ module.exports = (web3, options = {}) => {
         )
         .then(returnVersion(web3, getInfoTimeout))
     },
-    getVersionById (appId, versionId, getInfoTimeout) {
+    getVersionById(appId, versionId, getInfoTimeout) {
       return this.getRepository(appId)
-        .then((repository) =>
-          repository.methods.getByVersionId(versionId).call()
-        )
+        .then(repository => repository.methods.getByVersionId(versionId).call())
         .then(returnVersion(web3, getInfoTimeout))
     },
-    getLatestVersion (appId, getInfoTimeout) {
+    getLatestVersion(appId, getInfoTimeout) {
       return this.getRepository(appId)
-        .then((repository) =>
-          repository.methods.getLatest().call()
-        )
+        .then(repository => repository.methods.getLatest().call())
         .then(returnVersion(web3, getInfoTimeout))
     },
-    getLatestVersionContract (appId) {
+    getLatestVersionContract(appId) {
       return this.getRepository(appId)
-        .then((repository) =>
-          repository.methods.getLatest().call()
-        )
+        .then(repository => repository.methods.getLatest().call())
         .then(({ contractAddress }) => contractAddress)
     },
-    getLatestVersionForContract (appId, address, getInfoTimeout) {
+    getLatestVersionForContract(appId, address, getInfoTimeout) {
       return this.getRepository(appId)
-        .then((repository) =>
+        .then(repository =>
           repository.methods.getLatestForContractAddress(address).call()
         )
         .then(returnVersion(web3, getInfoTimeout))
     },
-    getAllVersions (appId) {
+    getAllVersions(appId) {
       return this.getRepository(appId)
-        .then((repository) =>
-          repository.methods.getVersionsCount().call()
-        )
-        .then((versionCount) => {
+        .then(repository => repository.methods.getVersionsCount().call())
+        .then(versionCount => {
           const versions = []
           for (let i = 1; i <= versionCount; i++) {
             versions.push(this.getVersionById(appId, i))
           }
           return Promise.all(versions)
         })
     },
-    isValidBump (appId, fromVersion, toVersion) {
-      return this.getRepository(appId)
-        .then((repo) => (
-          repo.methods.isValidBump(formatVersion(fromVersion), formatVersion(toVersion)).call()
-        ))
+    isValidBump(appId, fromVersion, toVersion) {
+      return this.getRepository(appId).then(repo =>
+        repo.methods
+          .isValidBump(formatVersion(fromVersion), formatVersion(toVersion))
+          .call()
+      )
+    },
+
+    /**
+     * Upload files under `directory` to storage provider `provider`.
+     *
+     * Returns a content URI string.
+     *
+     * @param {string} provider The name of an APM storage provider.
+     * @param {string} directory The directory that contains files to publish.
+     * @return {string} A string representing the contentURI
+     */
+    async uploadFilesToStorageProvider(provider, directory) {

```

Comment:
I don't really have anything against this, and it is the easiest solution, but it just feels like a bit of an abstraction leak.

An alternative way to achieve this is to add a hook into the publishing methods for an async "confirmation" check, that allows the caller to specify some custom logic once the files have been uploaded but before the publish happens. This would also allow the method to stay backwards compatible (if we really desired it, but no real reason to).

Reply:
You are right, I been thinking the same about the leak abstraction. 

On a second look I believe I'm able to get the `contentURI` from the intent something like: `params[4]` 🤔

In that case we don't need this PR at all 🙌